algorithm/include/gnuradio-4.0/algorithm/fourier/fft_common.hpp

```diff
@@ -6,33 +6,74 @@
+namespace gr::algorithm { namespace fft {
 
-template<ComplexType T, typename U>
-void
-computeMagnitudeSpectrum(const std::vector<T> &fftOut, std::vector<U> &magnitudeSpectrum, std::size_t fftSize, bool outputInDb) {
-    if (fftOut.size() < magnitudeSpectrum.size()) {
-        throw std::invalid_argument(fmt::format("FFT vector size ({}) must be more or equal than magnitude vector size ({}).", fftOut.size(), magnitudeSpectrum.size()));
+struct ConfigMagnitude {
+    bool computeHalfSpectrum = false;
+    bool outputInDb          = false;
+};
+
+template<std::ranges::input_range TContainerIn, std::ranges::output_range<typename TContainerIn::value_type::value_type> TContainerOut = std::vector<typename TContainerIn::value_type::value_type>,
+         typename T = TContainerIn::value_type>
+    requires(std::is_same_v<T, std::complex<float>> || std::is_same_v<T, std::complex<double>>)
+auto
+computeMagnitudeSpectrum(const TContainerIn &fftIn, TContainerOut &&magOut = {}, ConfigMagnitude config = {}) {
+    std::size_t magSize = config.computeHalfSpectrum ? fftIn.size() : (fftIn.size() / 2);
+    if constexpr (requires(std::size_t n) { magOut.resize(n); }) {
+        if (magOut.size() != magSize) {
+            magOut.resize(magSize);
+        }
+    } else {
+        static_assert(std::tuple_size_v<TContainerIn> == std::tuple_size_v<TContainerOut>, "Size mismatch for fixed-size container.");
     }
+
     using PrecisionType = typename T::value_type;
-    std::transform(fftOut.begin(), std::next(fftOut.begin(), static_cast<std::ptrdiff_t>(magnitudeSpectrum.size())), magnitudeSpectrum.begin(), [fftSize, outputInDb](const auto &c) {
-        const auto mag{ std::hypot(c.real(), c.imag()) * PrecisionType(2.0) / static_cast<PrecisionType>(fftSize) };
-        return static_cast<U>(outputInDb ? PrecisionType(20.) * std::log10(std::abs(mag)) : mag);
+    std::transform(fftIn.begin(), std::next(fftIn.begin(), static_cast<std::ptrdiff_t>(magSize)), magOut.begin(), [fftSize = fftIn.size(), outputInDb = config.outputInDb](const auto &c) {
+        const auto mag{ std::hypot(c.real(), c.imag()) * PrecisionType(2.) / static_cast<PrecisionType>(fftSize) };
+        return outputInDb ? PrecisionType(20.) * std::log10(std::abs(mag)) : mag;
     });
+
+    return magOut;
+}
+
+template<std::ranges::input_range TContainerIn, typename T = TContainerIn::value_type>
+    requires(std::is_same_v<T, std::complex<float>> || std::is_same_v<T, std::complex<double>>)
+auto
+computeMagnitudeSpectrum(const TContainerIn &fftIn, ConfigMagnitude config) {
+    return computeMagnitudeSpectrum(fftIn, {}, config);
 }
 
-template<ComplexType T, typename U>
-void
-computePhaseSpectrum(const std::vector<T> &fftOut, std::vector<U> &phaseSpectrum) {
-    if (fftOut.size() < phaseSpectrum.size()) {
-        throw std::invalid_argument(fmt::format("FFT vector size ({}) must be more or equal than phase vector size ({}).", fftOut.size(), phaseSpectrum.size()));
+struct ConfigPhase {
```

Comment:
I was initially hesitant when seeing an additional config type to `ConfigMagnitude` but the `ConfigPhase` definition here is warranted, since the `outputInDb` wouldn't make sense here. As a follow-up thought, what would make a lot of sense sense is to further specify:
 1. outputInDeg (default: false)
 2. unwrapPhase (default: false)
 
 As these are widespread post-processing options in that domain. Regarding naming: I am undecided whether specifying the prefix `outputXXX` is really necessary for brevity and context reasons, but I am OK if it is left as is.
 
**As a long-term goal*: we should look into, introduce and use physical unit safety. The [mp-units (v2)](https://github.com/mpusz/mp-units) is a very fine library, also for learning C++ meta-programming at its best. 
However, there are unresolved questions and challenges, particularly when it comes to:
 * interoperability w.r.t. runtime polymorphism (especially regarding type multiplicity because of additional quantity & unit NTTPs combinations)
 * interoperability w.r.t. complex data types (`std::simd`, `std::complex`, linear algebra, ...)

In my opinion, this would help constrain many of our block parameters and make GNU Radio in its various applications much safer to use also in industrial/critical applications.
Maybe a  leisurely discussion to be had for whether this is something for gnuradio-4.0 or rather 4.1. Maybe this could be done minimally invasive, with end-users do not having to pay a price for this. I think there is still a bit time as this would would need to be finalised only with the beta-release which I hope we can give to the wider GR community late spring (N.B. 'alpha' being what we want to push for the GR Architecture team and GR enthusiasts this autumn). Having this feature in-built could attract many commercial, and public-sponsored users and show-case GR & C++ at its best. For the time being, food for thought ...

@drslebedev, @wirew0rm, @ivan-cukic, @mattkretz, @mormj @jsallay